sandbox/windowed_spatial_avg.jl

```diff
@@ -0,0 +1,91 @@
+using Printf
+using Revise
+using Oceananigans
+using Oceananigans.Fields
+using Oceananigans.OutputWriters
+using Oceananigans.Advection
+using Oceananigans.Utils
+
+#++++ Model set-up
+Nx = Ny = Nz = 8
+Lx = Ly = Lz = 1
+N² = 1e-4 # s⁻²
+
+grid = RegularCartesianGrid(size=(Nx, Ny, Nz), extent=(Lx, Ly, Lz), topology=(Periodic, Periodic, Periodic))
+
+model = IncompressibleModel(
+                   grid = grid,
+              advection = UpwindBiasedFifthOrder(),
+            timestepper = :RungeKutta3,
+                closure = IsotropicDiffusivity(ν=1e-4, κ=1e-4),
+               coriolis = FPlane(f=1e-4),
+                tracers = (:b,), # P for Plankton
+               buoyancy = BuoyancyTracer(),
+)
+println()
+println(model)
+println()
+#-----
+
+#++++ Initial conditions
+u0(x, y, z) = y<Ly/2 ? 0.5 : 0.7
+v0(x, y, z) = 0.2
+w0(x, y, z) = 0.2*y
+
+set!(model, u=u0, v=v0, w=w0)
+#-----
+
+
+#++++
+struct WindowedSpatialAverage{F, S, D}
+          field :: F
+   field_slicer :: S
+           dims :: D
+end
+WindowedSpatialAverage(field; dims, field_slicer=FieldSlicer()) = WindowedSpatialAverage(field, field_slicer, dims)
+
+using Oceananigans.OutputWriters: slice_parent
+using Statistics: mean
+function (wsa::WindowedSpatialAverage)(model)
+    compute!(wsa.field)
+    window = slice_parent(wsa.field_slicer, wsa.field)
+    return dropdims(mean(window, dims=wsa.dims), dims=wsa.dims)
+end
+
+using NCDatasets: defVar
+using Oceananigans.Fields: reduced_location
+import Oceananigans.OutputWriters: xdim, ydim, zdim, define_output_variable!
+
+function define_output_variable!(dataset, wsa::WindowedSpatialAverage, name, array_type, compression, attributes, dimensions)
```

Comment:
I think you either want to dispatch on 

```julia
Union(WindowedSpatialAverage, WindowedTimeAverage{<:WindowedSpatialAverage})
```

or provide another implementation of `define_output_variable` for `output::WindowedTimeAverage{<:WindowedSpatialAverage}`.